### PR TITLE
[Fix #7728] Fix an error for `Style/OneLineConditional`

### LINKED ISSUE #7728: undefined method `prefix_not? in Style/OneLineConditional
## Expected behavior

When autocorrecting in Style/OneLineConditional you should not receive any errors

## Actual behavior

Autocorrect fails with:
```
undefined method `prefix_not?' for s(:self):RuboCop::AST::Node
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/style/one_line_conditional.rb:94:in `keyword_with_changed_precedence?'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/style/one_line_conditional.rb:82:in `requires_parentheses?'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/style/one_line_conditional.rb:74:in `expr_replacement'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/style/one_line_conditional.rb:67:in `to_ternary'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/style/one_line_conditional.rb:60:in `replacement'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/style/one_line_conditional.rb:39:in `block in autocorrect'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/cop.rb:36:in `call'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/corrector.rb:63:in `block (2 levels) in rewrite'
/usr/local/lib/ruby/gems/2.6.0/gems/parser-2.7.0.2/lib/parser/source/tree_rewriter.rb:220:in `transaction'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/corrector.rb:62:in `block in rewrite'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/corrector.rb:60:in `each'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/corrector.rb:60:in `rewrite'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/team.rb:136:in `autocorrect_all_cops'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/team.rb:72:in `autocorrect'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/team.rb:108:in `block in offenses'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/team.rb:125:in `investigate'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/team.rb:104:in `offenses'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cop/team.rb:44:in `inspect_file'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:296:in `inspect_file'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:244:in `block in do_inspection_loop'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:276:in `block in iterate_until_no_changes'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:269:in `loop'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:269:in `iterate_until_no_changes'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:240:in `do_inspection_loop'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:119:in `block in file_offenses'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:143:in `file_offense_cache'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:117:in `file_offenses'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:108:in `process_file'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:87:in `block in each_inspected_file'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:86:in `each'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:86:in `reduce'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:86:in `each_inspected_file'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:73:in `inspect_files'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/runner.rb:39:in `run'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli/command/execute_runner.rb:21:in `execute_runner'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli/command/execute_runner.rb:13:in `run'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli/command.rb:10:in `run'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli/environment.rb:17:in `run'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli.rb:65:in `run_command'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli.rb:72:in `execute_runners'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/lib/rubocop/cli.rb:41:in `run'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/exe/rubocop:13:in `block in <top (required)>'
/usr/local/Cellar/ruby/2.6.5/lib/ruby/2.6.0/benchmark.rb:308:in `realtime'
/usr/local/lib/ruby/gems/2.6.0/gems/rubocop-0.80.0/exe/rubocop:12:in `<top (required)>'
```

## Steps to reproduce the problem

Autocorrect this rather funky code:

```ruby
res.site (if self.is_a?(Chef::Resource::NoosferoSite) then self else self.site end)
```

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler. Here's an example:

```
0.80.0
```

### PR DESCRIPTION
**Pull request by koic**
Fixes #7728.

This PR fixes an error for `Style/OneLineConditional` when one of the branches contains a self keyword.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/ast/node/self_node.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    # A node extension for `self` nodes. This will be used in place of a
+    # plain node when the builder constructs the AST, making its methods
+    # available to all `self` nodes within RuboCop.
+    class SelfNode < Node
+      include MethodIdentifierPredicates
+      include ParameterizedNode
```
Comment (Drenmi) [inline lib/rubocop/ast/node/self_node.rb]: I don't think `self` can accept arguments. 🙂 

```
ruby-parse -e 'self(1)'
#=> error: unexpected token tLPAREN2
```

Comment (koic) [reply lib/rubocop/ast/node/self_node.rb]: Sure! I updated the PR.

File: lib/rubocop/ast/node/self_node.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+module RuboCop
+  module AST
+    # A node extension for `self` nodes. This will be used in place of a
+    # plain node when the builder constructs the AST, making its methods
+    # available to all `self` nodes within RuboCop.
+    class SelfNode < Node
+      include MethodIdentifierPredicates
+
+      # Always return `false` because `self` cannot have arguments.
+      def arguments?
```
Comment (Drenmi) [inline lib/rubocop/ast/node/self_node.rb]: Can you point me to the code that calls `#arguments?` on a `self` node? I think the call site probably should filter the relevant nodes instead. I can have a look. 🙂

Comment (koic) [reply lib/rubocop/ast/node/self_node.rb]: `SelfNode#arguments?` is called in the following code.
https://github.com/rubocop-hq/rubocop/blob/v0.80.0/lib/rubocop/cop/style/one_line_conditional.rb#L96

I was worried about choosing between implementing `SelfNode#arguments?` or guarding the above with `resopnd_to(:arguments?)`.

```diff
  def keyword_with_changed_precedence?(node)
    return false unless node.keyword?
    return true if node.prefix_not?

-   node.arguments? && !node.parenthesized_call?
+   node.respond_to?(:arguments?) && node.arguments? && !node.parenthesized_call?
  end
```

Comment (Drenmi) [reply lib/rubocop/ast/node/self_node.rb]: @koic Let me have a look. 🙂 Ideally we shouldn't do any of those. We should make sure incompatible nodes don't reach this code.

**Comment by tas50** [comment]
@koic Any chance this one can get merged so I can do testing on my side?

**Comment by koic** [comment]
@Drenmi Can we merge this PR to resolve the error first?

**Comment by Drenmi** [comment]
@koic Yes! Please go ahead. I don't have time to look into it for another couple of weeks or so. 🙂

**Comment by koic** [comment]
I get it. Thank you for taking your time!

**Pull request merged**